library/core/src/slice/index.rs

```diff
@@ -449,3 +451,113 @@ unsafe impl<T> SliceIndex<[T]> for ops::RangeToInclusive<usize> {
         (0..=self.end).index_mut(slice)
     }
 }
+
+/// Convert pair of `ops::Bound`s into `ops::Range` without performing any bounds checking and (in debug) overflow checking
```

Comment:
Somewhere in the user-facing documentation of the unchecked indexing methods, it should say that the user is also responsible for avoiding overflow when using an `Excluded` lower-bound or an `Included` upper-bound. Where does it say that?

Reply:
It seems like those could possibly go on the `impl` itself. There are other invariants documented on other `impl SliceIndex`es already.

Reply:
Isn't that basically undiscoverable?

Reply:
Yeh, it’s not as discoverable as directly on the methods themselves.

But it looks like this is actually also a problem for [`RangeInclusive::get_unchecked`](https://github.com/rust-lang/rust/blob/89fdb30892dbe330730ad1a1c1fe45b9046c2973/library/core/src/slice/index.rs#L390), which doesn’t appear to document the requirement that `self.end != usize::MAX` in the public API docs.

Maybe we should flesh out the docs for `[T]::get_unchecked{_mut}` a bit to mention the possibility of overflowing bounds if they’re converted between inclusive/exclusive?

Reply:
Yes, I agree this is an existing problem.

> Maybe we should flesh out the docs for [T]::get_unchecked{_mut} a bit to mention the possibility of overflowing bounds if they’re converted between inclusive/exclusive?

Sounds reasonable.

Reply:
The documentation right now says:

> Calling this method with an out-of-bounds index or a dangling slice pointer is undefined behavior even if the resulting reference is not used.

Since `usize::MAX` would always be out of bounds for anything taking up space, we're only talking about overflows in `[T]::get_unchecked` for ZSTs here, right? Or am I missing something?

Reply:
And even for ZSTs, slices can not be longer than usize::MAX. So an `Included(usize::MAX)` bound already counts as out-of-bounds, permitting UB.

Reply:
> So an Included(usize::MAX) range end already counts as out-of-bounds, permitting UB.

Hm, I can agree with that interpretation, but I feel it would be worth calling out explicitly that if there is an overflow while computing the bounds, that also counts as out-of-bounds.

Reply:
At the very least it would be a great inline comment here and in `RangeInclusive`.